test/tools/cli/IonProcess.java

```diff
@@ -387,57 +361,101 @@ private static void embeddedEventToIon(IonReader ionReader,
-                                .createIonWriterWithImports(outputStream, symbolTables);
+                        handleSymbolTableEvent(embeddedContext, event, outputStream, args);
                     }
                 } while (ionReader.next() != null);
 
                 if (out.length() > 0) {
-                    tempWriter.finish();
-                    ionWriterForOutput.writeString(out.toString());
+                    embeddedContext.getIonWriter().finish();
+                    processContext.getIonWriter().writeString(out.toString());
+                }
+            } finally {
+                IonWriter ionWriter = embeddedContext.getIonWriter();
+                if (ionWriter != null) {
+                    try {
+                        ionWriter.close();
+                    } catch (IOException e) {
+                        System.err.println(e.getMessage());
+                    }
                 }
             }
             if (finish) break;
         }
-        ionWriterForOutput.stepOut();
+        processContext.getIonWriter().stepOut();
+    }
+
+    private static void handleSymbolTableEvent(ProcessContext processContext,
+                                               Event event,
+                                               OutputStream outputStream,
+                                               ProcessArgs args) throws IOException {
+        processContext.getIonWriter().close();
+        ImportDescriptor[] imports = event.getImports();
+        SymbolTable[] symbolTables = new SymbolTable[imports.length];
+        for (int i = 0; i < imports.length; i++) {
+            SymbolTable symbolTable = ION_SYSTEM.newSharedSymbolTable(
+                    imports[i].getImportName(), imports[i].getVersion(), null);
+            symbolTables[i] = symbolTable;
+        }
+        processContext.setIonWriter(
+                args.getOutputFormat().createIonWriterWithImports(outputStream, symbolTables));
     }
 
     //
     //  helper functions
     //
 
-    private static void writeIonByType(Event event, IonWriter ionWriter) throws IOException {
+    private static void writeIonByType(Event event, IonWriter ionWriter) {
         setAnnotationAndField(event, ionWriter);
-        event.getValue().writeTo(ionWriter);
+        IonValue value = event.getValue();
+        value.setTypeAnnotationSymbols(event.getAnnotations());
+        value.writeTo(ionWriter);
     }
 
-    private static Event eventStreamToEvent(IonReader ionReader) {
-        Event event = new Event();
-        ionReader.stepIn();
+    private static void setAnnotationAndField(Event event, IonWriter ionWriter) {
+        SymbolToken field = event.getFieldName();
+        SymbolToken[] annotations = event.getAnnotations();
+        if (field != null) {
+            if (!ionWriter.isInStruct()) throw new IonException("invalid field_name inside STRUCT");
+            ionWriter.setFieldNameSymbol(field);
+        } else if (ionWriter.isInStruct()) {
```

Comment:
Should this be an error? Inside a struct, every value *must* have a field name symbol token.